src/vm/ilmarshalers.h

```diff
@@ -602,7 +602,7 @@ class ILMarshaler
                 nativeSize = wNativeSize;
             }
 
-#if defined(_TARGET_X86_) || (defined(_TARGET_AMD64_) && defined(_WIN32))
+#if defined(_TARGET_X86_) || ((defined(_TARGET_AMD64_) || defined(_TARGET_ARM64_)) && defined(_WIN32))
```

Comment:
Have you checked the behavior on Windows ARM? I would bet that `ARM` should be in this condition as well.

Reply:
Also, you may check the behavior on Linux x86 - I would expect it to be in sync with other Linux platforms and not needing this. This would make this ifdef:

`#ifdef PLATFORM_WINDOWS`

Reply:
I just went and verified. The current behavior (without this block) works perfectly on Windows ARM. Windows ARM doesn't need this condition.

I'm working to check Linux x86 now.

Reply:
I've been testing with the `ThisCallNative` unit test that I added that's at https://github.com/dotnet/coreclr/blob/master/tests/src/Interop/PInvoke/Miscellaneous/ThisCall/ThisCallNative.cpp

It's possible that the JIT is doing the right thing here on ARM so the interop space isn't having to account for the return buffer itself.

Reply:
It also works with the following (just tested):

```

struct SizeF
{
    float width;
};

class C
{
    int dummy = 0xcccccccc;
    float width;

public:
    C(float width, float height)
        :width(width)
    {}

    virtual SizeF GetSize()
    {
        return {width};
    }
};

```

Reply:
Tried out your test case and it failed. Looks like the JIT isn't correctly covering this case.

I'll enable this on all Windows platforms and move the Unix x86 part into a separate block for cleanliness until I can verify that it isn't needed.

Reply:
And does it work with `struct X { int x; };` that I have shared above?

Reply:
`float`s are HFAs on ARM and other platforms; so they tend to follow different rules when it comes to calling conventions.

Reply:
Looks like HFAs need to be handled separately. They can't use the byref return on Windows ARM32. I should be able to disable that relatively easily.

Reply:
The Unix code is still needed (just tested it locally).